airflow/jobs/scheduler_job_runner.py

```diff
@@ -2113,44 +2114,58 @@ def _activate_referenced_assets(assets: Collection[AssetModel], *, session: Sess
         active_name_to_uri: dict[str, str] = {name: uri for name, uri in active_assets}
         active_uri_to_name: dict[str, str] = {uri: name for name, uri in active_assets}
 
-        def _generate_dag_warnings(offending: AssetModel, attr: str, value: str) -> Iterator[DagWarning]:
+        def _generate_warning_message(
+            offending: AssetModel, attr: str, value: str
+        ) -> Iterator[tuple[str, str]]:
             for ref in itertools.chain(offending.consuming_dags, offending.producing_tasks):
-                yield DagWarning(
-                    dag_id=ref.dag_id,
-                    error_type=DagWarningType.ASSET_CONFLICT,
-                    message=f"Cannot activate asset {offending}; {attr} is already associated to {value!r}",
+                yield (
+                    ref.dag_id,
+                    f"Cannot activate asset {offending}; {attr} is already associated to {value!r}",
                 )
 
-        def _activate_assets_generate_warnings() -> Iterator[DagWarning]:
+        def _activate_assets_generate_warnings() -> Iterator[tuple[str, str]]:
             incoming_name_to_uri: dict[str, str] = {}
             incoming_uri_to_name: dict[str, str] = {}
             for asset in assets:
                 if (asset.name, asset.uri) in active_assets:
                     continue
                 existing_uri = active_name_to_uri.get(asset.name) or incoming_name_to_uri.get(asset.name)
                 if existing_uri is not None and existing_uri != asset.uri:
-                    yield from _generate_dag_warnings(asset, "name", existing_uri)
+                    yield from _generate_warning_message(asset, "name", existing_uri)
                     continue
                 existing_name = active_uri_to_name.get(asset.uri) or incoming_uri_to_name.get(asset.uri)
                 if existing_name is not None and existing_name != asset.name:
-                    yield from _generate_dag_warnings(asset, "uri", existing_name)
+                    yield from _generate_warning_message(asset, "uri", existing_name)
                     continue
                 incoming_name_to_uri[asset.name] = asset.uri
                 incoming_uri_to_name[asset.uri] = asset.name
                 session.add(AssetActive.for_asset(asset))
 
-        warnings_to_have = {w.dag_id: w for w in _activate_assets_generate_warnings()}
+        def _get_first_item(x: Sequence[Any]) -> Any:
+            return x[0]
+
+        warnings_to_have = {
```

Comment:
@uranusjr, I'm not sure whether we want to warn a case like this 🤔 It kinda makes the logic a bit more complicated. The message is just concatenated.

```python
    schedule=(
        Asset(name="asset1_producer", uri="s3://bucket/asset1_producer1")
        | Asset(name="asset1_producer", uri="s3://bucket/asset1_producer2")
        | Asset(name="asset1_producer", uri="s3://bucket/asset1_producer3")
        | Asset(name="asset1_producer", uri="s3://bucket/asset1_producer4")
    ),
```

Reply:
Do you mean we should reduce the warnings from 3 to 1?

Reply:
In the existing logic, we only raise the latest one (if I'm not mistaken). In this implementation, the error will look like "Cannot activate asset Asset(...); ... is already associated to ...\nCannot activate asset Asset(...); ... is already associated to ...".

I don't think we can add multiple DagWarning unless we want to remove the unique constriant 🤔